### PR TITLE
Feat/controller implementation

### PR DESCRIPTION
**Pull request by Courtcircuits**
Implementation of the controller

### CODE REVIEW

File: .idea/.gitignore
Comment (sameo) [inline .idea/.gitignore:1]: Please remove all `.idea`files.

File: controller/src/action/action_repository.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+use sqlx::PgPool;
+use std::sync::Arc;
+
+use crate::parser::pipe_parser::Type;
+
+pub struct Action {
+    pub id: i64,
+    pub pipeline_id: i64,
+    pub name: String,
+    container_uri: String,
```
Comment (hugoponthieu) [inline controller/src/action/action_repository.rs:10]: What is container_uri ?

**Review by hugoponthieu — CHANGES_REQUESTED** [review]
I was wondering if adding route for fetching a stream that expose the state of the pipeline was possible

File: docs/arch/lexicon.md [hunk 2]
```diff
@@ -7,4 +6,3 @@
 - **Agent pool**: the set of all registered agents. It is a scheduler's entire knowledge about available computing resources.
 - **Pipeline**: a set of actions to be executed. It is declared as a YAML file.
 - **Scheduling**: selection of an agent on which to execute an action.
-
```
Comment (theotchlx) [inline docs/arch/lexicon.md:10]: Please re-add this removed empty line at the end of this .md file

No blank lines at the end of a file is not up to markdown standards,

And generally, it is best to leave one blank line at the end of any file (source code included), as many older tools won't work properly if a line feed isn't present. Also it is a POSIX standard, as it is the definition of a line: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

**Review by theotchlx — CHANGES_REQUESTED** [review]
Hi, thank you for this great work and PR.

I suggested below some adjustments and questions

File: controller/src/scheduler/mod.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            .message()
+            .await
+            .map_err(|_err| PipelineServiceError::SchedulerError)?
+        {
+            println!("RESPONSE={:?}", response);
+        }
+
+        Ok(())
+    }
+}
```
[the remark this answers is not in the document]
Comment (Courtcircuits) [reply controller/src/scheduler/mod.rs:56]: You can put code inside `mod.rs`. E.g : https://github.com/tokio-rs/axum/blob/main/axum/src/response/mod.rs

File: controller/src/pipeline/pipeline_service.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        &self,
+        _pipeline: Pipeline,
+        _repo_url: String,
+    ) -> Result<(), PipelineServiceError> {
+        let _client = Arc::clone(&self.client);
+        // for action in pipeline.actions {
+        //     info!("Sending action: {:?}", action);
+        //     self.send_action(client.clone(), Arc::new(action), repo_url.clone())
+        //         .await?;
+        // }
```
[the remark this answers is not in the document]
Comment (Courtcircuits) [reply controller/src/pipeline/pipeline_service.rs:113]: Nope we will get rid of it

File: controller/src/pipeline/pipeline_controller.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                }
+            } else {
+                info!("Error while creating pipeline");
+                return HttpResponse::InternalServerError().finish();
+            }
+
+            HttpResponse::Ok().finish()
+        }
+        Err(ParsingError::YamlNotCompliant) => HttpResponse::BadRequest().body("Invalid yaml"),
+        Err(err) => HttpResponse::BadRequest().body(format!("{:?}", err)), //TODO: replace this by exhaustive match
```
[the remark this answers is not in the document]
Comment (Courtcircuits) [reply controller/src/pipeline/pipeline_controller.rs:74]: Yes

File: controller/.gitignore
[the remark this answers is not in the document]
Comment (Courtcircuits) [reply controller/.gitignore:1]: We'll do it

**Review by hugoponthieu — APPROVED** [review]
The code is clean, consider merging this part

**Comment by Courtcircuits** [comment]
> The code is clean, consider merging this part

We still need to separate the "commands" into a dedicated repository. This is in progress, so until it's committed I would rather not merge this branch. Sorry for the wait

**Pull request merged**